token/src/token.rs

```diff
@@ -0,0 +1,89 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::scalar::Scalar;
+use zkvm::*;
+
+pub struct Token {
+    flavor: Predicate,
+    metadata: Vec<u8>,
+    nonce: Scalar,
+}
+
+impl Token {
+    pub fn new(pred: Predicate, metadata: &[u8]) -> Self {
+        Token {
+            flavor: pred,
+            metadata: metadata.to_vec(),
+            nonce: Scalar::random(&mut rand::thread_rng()),
+        }
+    }
+
+    pub fn issue<'a>(&self, program: &'a mut Program, qty: u64) -> &'a mut Program {
+        let nonce = Predicate::Key(VerificationKey::from_secret(&self.nonce));
+        program
+            .push(Commitment::blinded(qty)) // stack: qty
+            .var() // stack: qty-var
+            .push(Commitment::unblinded(Value::issue_flavor(
+                &self.flavor,
+                Data::Opaque(self.metadata.to_vec()),
```

Comment:
oh, i overlooked. I don't think we need to ask for `Data` in `issue_flavor`. We only need a `Vec<u8>`.